pkg/drivers/sqlite/sqlite.go

```diff
@@ -77,172 +65,68 @@ func NewVariant(ctx context.Context, driverName, dataSourceName string) (server.
-				lease INTEGER,
-				value BLOB,
-				old_value BLOB
-			)`
-
-	_, err := txn.Exec(createTableSQL)
-	if err != nil {
+	row := db.QueryRowContext(ctx, `PRAGMA user_version`)
+	if err := row.Scan(&schemaVersion); err != nil {
 		return err
 	}
 
-	return nil
-}
+	if schemaVersion == databaseSchemaVersion {
+		return nil
+	}
 
-// removeTable drops the table with the given name, in the given transaction.
-func removeTable(txn *sql.Tx, tableName string) error {
-	dropTableSQL := fmt.Sprintf(`DROP TABLE IF EXISTS %s`, tableName)
-	_, err := txn.Exec(dropTableSQL)
+	txn, err := db.BeginTx(ctx, nil)
 	if err != nil {
 		return err
 	}
+	defer txn.Rollback()
 
-	return nil
-}
-
-// alterIndices drops the given old database indices and creates
-// the given new ones.
-func alterIndices(txn *sql.Tx, d *generic.Generic) error {
-	if d.LockWrites {
-		d.Lock()
-		defer d.Unlock()
-	}
-
-	dropIndices := []string{
-		`DROP INDEX IF EXISTS kine_name_index`,
-		`DROP INDEX IF EXISTS kine_name_prev_revision_uindex`,
-	}
-
-	// Drop the old indices
-	for _, stmt := range dropIndices {
-		_, err := txn.Exec(stmt)
-		if err != nil {
-			return err
-		}
-	}
-
-	createIndices := []string{
-		`CREATE INDEX IF NOT EXISTS kine_name_index ON kine (name, id)`,
-		`CREATE UNIQUE INDEX IF NOT EXISTS kine_name_prev_revision_uindex ON kine (prev_revision, name)`,
-	}
-
-	// Create the new indices
-	for _, stmt := range createIndices {
-		_, err := txn.Exec(stmt)
-		if err != nil {
-			return err
+	for i := 0; i < retryAttempts; i++ {
+		if err := migrate(ctx, txn); err != nil {
+			if sqliteErr, ok := err.(sqlite3.Error); ok && sqliteErr.ExtendedCode == sqlite3.ErrBusySnapshot {
+				// Another write transaction got in the way (SERIALIZABLE isolation).
+				// The whole transaction can be retried.
+				continue
+			}
+			return errors.Wrap(err, "migration failed")
 		}
+		break
 	}
 
-	return nil
-}
-
-// countTable counts the number of tables with the given name, in the given transaction.
-func hasTable(ctx context.Context, txn *sql.Tx, tableName string) (bool, error) {
-	tableListSQL := fmt.Sprintf(`SELECT COUNT(*) FROM sqlite_master WHERE type = 'table' AND name = '%s'`, tableName)
-	row := txn.QueryRowContext(ctx, tableListSQL)
-	var tableCount int
-	if err := row.Scan(&tableCount); err != nil {
-		return false, err
-	}
-
-	return tableCount != 0, nil
+	return txn.Commit()
 }
 
-// migration copies rows of the old key value table to the kine
-// table ONLY IF the key value table exists and migration has not been
-// done already. It creates the kine table if it doesn't exist.
-func migration(ctx context.Context, txn *sql.Tx, d *generic.Generic) error {
-	userVersionSQL := `PRAGMA user_version`
-	row := txn.QueryRowContext(ctx, userVersionSQL)
-
+// migrate tries to migrate from a version of the database
+// to the target one.
+func migrate(ctx context.Context, txn *sql.Tx) error {
 	var userVersion int
+
+	row := txn.QueryRowContext(ctx, `PRAGMA user_version`)
```

Comment:
Maybe we could pass the schemaVersion as parameter to `migrate()` to avoid the second pragma call for `user_version` ?